Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionDynamicScriptsCocoa.mm

```diff
@@ -0,0 +1,114 @@
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#if !__has_feature(objc_arc)
+#error This file requires ARC. Add the "-fobjc-arc" compiler flag for this file.
+#endif
+
+
+#import "config.h"
+#import "WebExtensionDynamicScripts.h"
+
+#if ENABLE(WK_WEB_EXTENSIONS)
+#import "WKContentWorld.h"
+#import "WKFrameInfoPrivate.h"
+#import "WKWebViewPrivate.h"
+#import "WebExtensionContext.h"
+#import "WebExtensionUtilities.h"
+#import "_WKFrameHandle.h"
+#import "_WKFrameTreeNode.h"
+#import <wtf/URL.h>
+
+namespace WebKit {
+
+namespace DynamicScripts {
+
+void addResultAndCompleteIfDone(id resultOfJSExecution, WKFrameInfo *frame, NSString *errorMessage, size_t numberOfExecutions, Vector<WebExtensionScriptInjectionResultParameters>&& results, CompletionHandler<void(std::optional<Vector<WebExtensionScriptInjectionResultParameters>>)>&& callback)
+{
+    WebExtensionScriptInjectionResultParameters result;
+    result.result = resultOfJSExecution ?: @"";
+    result.frameID = frame.isMainFrame ? mainFrameID : frame._handle.frameID;
+    if (errorMessage.length)
+        result.error = errorMessage;
+
+    results.append(result);
+    if (results.size() == numberOfExecutions)
+        callback(results);
+}
+
+void populateFrames(_WKFrameTreeNode *currentNode, size_t startingFrameID, bool allFrames, bool foundStartingFrame, Vector<RetainPtr<_WKFrameTreeNode>>&& frames)
+{
+    if (foundStartingFrame)
+        frames.append(WTFMove(currentNode));
+
+    WKFrameInfo *info = currentNode.info;
+    size_t currentFrameID = info.isMainFrame ? mainFrameID : info._handle.frameID;
+    if (currentFrameID == startingFrameID) {
+        frames.append(WTFMove(currentNode));
+        if (!allFrames)
+            return;
+
+        foundStartingFrame = true;
+    }
+
+    for (_WKFrameTreeNode *child in currentNode.childFrames)
+        DynamicScripts::populateFrames(child, startingFrameID, allFrames, foundStartingFrame, WTFMove(frames));
+}
+
+void populateFrames(_WKFrameTreeNode *currentNode, std::optional<Vector<size_t>> frameIDs, Vector<RetainPtr<_WKFrameTreeNode>>&& frames)
+{
+    WKFrameInfo *info = currentNode.info;
+    size_t currentFrameID = info.isMainFrame ? mainFrameID : info._handle.frameID;
+    if (frameIDs && frameIDs->contains(currentFrameID))
+        frames.append(WTFMove(currentNode));
```

Comment:
I don't think you should have the `WTFMove` here. At best it's redundant, at worst it actually causes the `currentNode` pointer to be cleared.